jupyter_client/restarter.py

```diff
@@ -31,6 +34,12 @@ class KernelRestarter(LoggingConfigurable):
 
     time_to_dead = Float(3.0, config=True, help="""Kernel heartbeat interval in seconds.""")
 
+    stable_start_time = Float(
+        10.0,
```

Comment:
Since there isn't a way to _cancel_ this, I think `5.0` may be a more reasonable default since most local kernels startup well within this time.

I suppose if a restart is necessary during the stable-start-time window, we'll just increment the restart count (since `restarting` is still True).  In essence, this is essentially extending our ability to restart (possibly with fresh ports if `iniitial_startup` is True) for the duration of the stable start time - is that correct?

Reply:
I'm not sure what you mean by "cancel" here.

Separately, these were the arguments that made me want to err on the side of a higher default value:

The main disadvantage that I see to having a high value for this trait would be e.g. if a client executes some notebook cells within this timeframe that causes the kernel to terminate unexpectedly. In that case, the new termination would be counted together with the previous restarts, even if the kernel strictly speaking finished starting up.

The main disadvantage for having the value too low is a risk of an infinite restart loop when there is a port conflict:
1. During initial startup, there is a port conflict that causes the kernel to die.
2. If a `poll` is triggered after the `stable_start_time` but before the kernel dies, the startup will be considered ok, and the kernel will no longer be considered restarting (if it was).
3. When the death of the kernel is subsequently detected, it will be restarted keeping the same ports, and with a fresh restart sequence (attempt no 1). Go to 2.
In this case, you can escape the loop if the timings of the next attempt are different (the use of `run_sync(self.poll)` ensures we now have that possibility), but if the kernel start up is consistently higher than `stable_start_time + time_to_dead` it is likely to keep going..

Reply:
> In essence, this is essentially extending our ability to restart (possibly with fresh ports if iniitial_startup is True) for the duration of the stable start time - is that correct?

Yes, that is correct. Said differently, if the kernel dies within stable-start-time, we consider it part of the same "restart-sequence" (and if the initial startup is part of that sequence, we will keep changing ports).

Reply:
Thanks for the more detailed explanation - that all makes sense.  Would you be able to include this explanation in a comment for future maintainers (or reference the PR)?

> I'm not sure what you mean by "cancel" here.

I meant the ability to indicate that a stable start has been completed prior to the timeout.  However, as you said, there's really no harm in letting subsequent restarts (including those following stable starts) from participating in the stable-start-initial-startup behavior.  I guess the downside is you may not get your 5 restart attempts, but I've always found that if the first restart doesn't work, then it's likely that none of the other restarts will succeed either and the limit will be reached anyway.

Reply:
I had written a comment in the commit message, but I'll add something in the code as well, as I appreciate that not everyone uses git-blame all the time 😄